Partial support for encrypted audio in k573dio (Konami System 573 Digital I/O)

windyfairy opened the pull request:
This patch adds partial high level support for the FPGA and MP3 subboard in k573dio. 

### MP3 decoder
minimp3 was chosen for MP3 decoding because it was easy to integrate, easy to modify, and is licensed under CC0, which is compatible with GPL.

### Compatibility
All Dance Dance Revolution/Dancing Stage games except the Solo series (ddrs2k, ddrs2kj, ddrsbm) seem to be working based on my testing.

The Guitar Freaks/Drummania series also seems to work up until GF10DM9, at which point there are some playback issues in-game. The MPEG control flag must be set during the song previews, menu BGMs, etc, but if you set it during the actual song then it stops the chart from progressing. I can't find any way to differentiate the two scenarios.

Mambo a Gogo works.

Dance Maniax series *seemingly* works, but there seems to be an issue with the underlying PSX audio emulation which causes it to play a buzzing sound over top everything. This is in base MAME and not an issue with this PR.

Martial Beat is not possible to test due to an I/O error at boot.

Playback is rough in some areas. The MPEG control flag in particular seems to be quite temperamental, and getting that right is the key to making the currently broken games work. I've made a hardware test which I ran on the real Sytem 573 Digital I/O hardware, and during MP3 playback the MPEG control flag always returns 0x1000, and when it's not playing anything it returns 0x0000, but properly emulating that behavior does not seem to fix things. There is possible some timing issues involved which seem to be beyond my capabilities and knowledge of how the hardware is managing it. 

The audio decryption code for all games is working, and with some very hacky workarounds for the MPEG control flag it's possible to get the currently broken stuff working as well.

### Assistance/advice requested

- Anything involving the MPEG control flag. I think there may be 3 or 4 different versions of the FPGA firmware from what I could tell by looking at the game's data, so this may possibly be a result of different behavior between the FPGA firmwares.
- The MP3 decoding loop. I'm not sure at what timing the game actually behinds decoding things, and I feel like my current implementation is just not very good but I'm not sure what to do about it.
- Ideally the MAS3507D code would be used more often I think, since it's kinda just being worked around for now.
- I'm iffy about modifying the samples device code like I am doing here. I'm especially iffy about the update callback which may have unintended performance implications in other areas. Is there a better way to do this?
- I've made some modifications to minimp3 but I'd like to revisit this and try to make it work with a non-modified minimp3 in the future if possible, so it's easier to keep the library updated when needed.
- Buffer speed workaround. It works perfectly fine from what I've tested, but I feel like this can be done better. It may have something to do with the bitrate of the MP3 or something along those lines.

rb6502 commented:
I haven't looked at this in depth yet, but I know we're gonna want it to use our homegrown MPEG decoder in src\devices\machine\mpeg_audio.cpp.  It may or may not actually support MP3 yet, but that can be added.

windyfairy commented:
@rb6502 Understood. I was afraid that would be the case, and currently mpeg_audio only supports layer 2 and not layer 3 from what I was able to figure out. Unfortunately audio codecs are out of my comfort zone which is why I decided to use minimp3 in order to offload that work and focus on everything else.

ajrhacker commented:
It's also expected that third-party libraries used by MAME be placed in the `3rdparty` directory rather than `src/devices/sound`.

windyfairy commented:
@ajrhacker Thanks, I didn't know that. I've moved it to the `3rdparty` folder and added the license information to the `README.md`. :+1:

galibert commented:
I have things with higher priority than layer 3 support (which is roughly
70% done fwiw), so I don't mind accepting that change and getting rid of
the library later.

  OG.


On Tue, May 14, 2019 at 9:21 PM R. Belmont <notifications@github.com> wrote:

> I haven't looked at this in depth yet, but I know we're gonna want it to
> use our homegrown MPEG decoder in src\devices\machine\mpeg_audio.cpp. It
> may or may not actually support MP3 yet, but that can be added.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mamedev/mame/pull/5055?email_source=notifications&email_token=ACGSF4KVIUG326D7DMYCZLTPVMGJ5A5CNFSM4HMZ6O5KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVMQRBQ#issuecomment-492374150>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACGSF4ONQVCBABE5THJHJLTPVMGJ5ANCNFSM4HMZ6O5A>
> .
>

galibert commented on src/devices/sound/mas3507d.cpp:
```diff
@@ -226,14 +252,151 @@ void mas3507d_device::i2c_device_got_stop()
 	logerror("MAS I2C: got stop\n");
 }
 
+int gain_to_db(int val) {
```
Urgh, that's a direct multiplier.  In other words, db = 20*log10((0x100000 - v)/0x80000)

galibert commented on src/mame/machine/k573enc.h:
```diff
@@ -0,0 +1,120 @@
+
+#include <map>
+
+std::unordered_map<uint16_t, uint32_t> k573enc_lookup({
```
What is that table supposed to be?  There is no such thing in the fpga.

windyfairy replied:
This ties directly into the other comment so I'll respond to both here.

>Do you mean that ddrsbm has a different fpga firmware?

ddrsbm is the only game on the k573dio platform that uses a different encryption algorithm compared to everything else. The fake FPGA algorithm is an approximation of the FPGA algorithm that, when given a decryption key and scramble key and counter (ddrsbm doesn't use a counter though), you can generate the same decrypted data that the real algorithm would provide.

ddrsbm is also the only game that uses FPGA key 1 only as its input. Every other game uses key parts 2 and 3. Its scramble key is also always derived from the decryption key, whereas the scramble key would need to be something completely different from the decryption key for every other game's data.

The table itself was derived using real hardware to get the plain text data using the real FPGA firmware, and then calculating the decryption key based on the plaintext data. With the lack of the real algorithm being run on the FPGA for this specific game, the fake algorithm provides the same results but requires these hardcoded keys.

galibert commented on src/mame/machine/k573fpga.cpp:
```diff
@@ -0,0 +1,592 @@
+    return (value >> n) & 1;
+}
+
+inline uint16_t BitSwap16(int v, int b15, int b14, int b13, int b12, int b11, int b10, int b9, int b8,
```
We have templates in MAME for that, please use them.  I'm referring to BIT, BITSWAP16, etc.

windyfairy replied:
Thank you. I found the functions in the utils class and made the appropriate changes and confirmed everything still works. :+1:

galibert commented on src/mame/machine/k573fpga.cpp:
```diff
@@ -0,0 +1,592 @@
+inline uint16_t k573fpga_device::fpga_decrypt_byte_fake(uint16_t data, uint32_t crypto_idx)
+{
+    // Not the real algorithm. Only used for DDR Solo Bass Mix (ddrsbm)
+    int32_t crypto_enc_idx = find_enc_key();
```
Do you mean that ddrsbm has a different fpga firmware?

windyfairy replied:
I've elaborated in [this comment](https://github.com/mamedev/mame/pull/5055#discussion_r284123984), but yes, ddrsbm uses a completely different FPGA firmware from everything else. Every other game seems to use an FPGA firmware derived at the very least from a standard program, and then ddrsbm comes in and does something completely different.

galibert commented on src/mame/machine/k573fpga.cpp:
```diff
@@ -0,0 +1,592 @@
+
+    int new_samples = 0;
+
+    if (!use_fake_fpga && mp3_start_adr != mp3_dynamic_base) {
```
Eventually that will be fully streamed and most of that code will just go away, but that will require a working mpeg_audio.cpp.

galibert commented:
Ok, I had checked at the time that every game used the same firmware, but
ddrsbm happened since.  Let's see if I can reverse that one...

  OG.


On Wed, May 15, 2019 at 9:45 AM Windy Fairy <notifications@github.com>
wrote:

> *@windyfairy* commented on this pull request.
> ------------------------------
>
> In src/mame/machine/k573fpga.cpp
> <https://github.com/mamedev/mame/pull/5055#discussion_r284124502>:
>
> > +
> +    v ^= BitSwap16(
> +        crypto_key3,
> +        7, 0, 6, 1,
> +        5, 2, 4, 3,
> +        3, 4, 2, 5,
> +        1, 6, 0, 7
> +    );
> +
> +    return (v >> 8) | ((v & 0xff) << 8);
> +}
> +
> +inline uint16_t k573fpga_device::fpga_decrypt_byte_fake(uint16_t data, uint32_t crypto_idx)
> +{
> +    // Not the real algorithm. Only used for DDR Solo Bass Mix (ddrsbm)
> +    int32_t crypto_enc_idx = find_enc_key();
>
> I've elaborated in this comment
> <https://github.com/mamedev/mame/pull/5055#discussion_r284123984>, but
> yes, ddrsbm uses a completely different FPGA firmware from everything else.
> Every other game seems to use an FPGA firmware derived at the very least
> from a standard program, and then ddrsbm comes in and does something
> completely different.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/mamedev/mame/pull/5055?email_source=notifications&email_token=ACGSF4IUU533NME7KK6GB2LPVO5SNA5CNFSM4HMZ6O5KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBYVFLNI#discussion_r284124502>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACGSF4PUCAM647XVLX3DN6DPVO5SNANCNFSM4HMZ6O5A>
> .
>

rb6502 commented:
Alright.  So we're just waiting on the request gain_to_db() to turn it into math instead of a big ugly switch and this will be good to go in.

For future reference, we do have an I2C bus implementation and I2C devices so you didn't need to reinvent that wheel, but I'm not gonna hold this up for that.

windyfairy commented:
@rb6502 @galibert 
I've replaced the big switch with the math provided by @galibert. I wrote a test case on my end to verify that all of the values are returning the expected results, and everything passed with the exception of the already known odd case of -98 and -97 sharing the same gain value.

I also fixed a regression that resulted in MP3s that should've been decoded fully in one shot try to (unsuccessfully) look for more audio when the next buffer sync came around near the end of the MP3 playback. This was due to the last decoded address was not being updated to include the amount of bytes that were decoded in that specific case.

Please give it a look over when you guys have the chance. :bowing_man:

Tafoid commented:
Checking it this morning I noticed that all affected k573sys machines have a sample requirement and are expected a file with "audio.wav" inside.  CLRMAME, too, sees the k573_dio device and expects this same .wav sample file.

Perhaps the device was added incorrectly or something else can be done to fix this?

For example:
>c:\mamegit\mame>mame64 dncfrks -listsamples
Samples required for driver "dncfrks".
audio

smf- commented:
I've let him know it needs converting to device_sound_interface

rb6502 commented:
We can convert it when OG gets his decoder online.  I didn't want to convert it now and then have to throw away the work.
